lib/src/internal/http/http_handler.dart

```diff
@@ -1,84 +1,108 @@
-
-  final List<HttpBucket> _buckets = [];
-  late final HttpBucket _noRateBucket;
-
   late final http.Client httpClient;
 
   final Logger logger = Logger("Http");
   final INyxxRest client;
 
+  RestEventController get _events => client.eventsRest as RestEventController;
+
+  final Map<String, HttpBucket?> _bucketByRequestRateLimitId = <String, HttpBucket?>{};
+  DateTime globalRateLimitReset = DateTime.fromMillisecondsSinceEpoch(0);
+
   /// Creates an instance of [HttpHandler]
   HttpHandler(this.client) {
-    _noRateBucket = HttpBucket("", this);
     httpClient = http.Client();
   }
 
-  Future<HttpResponse> execute(HttpRequest request) async {
-    if (request.auth) {
-      request.headers.addAll({"Authorization": "Bot ${client.token}"});
+  HttpBucket? _upsertBucket(HttpRequest request, http.StreamedResponse response) {
+    //Get or Create Bucket
+    HttpBucket? bucket = _bucketByRequestRateLimitId.values.toList().firstWhereSafe((bucket) => bucket?.isInBucket(response) ?? false) ?? HttpBucket.fromResponseSafe(response);
+    //Update Bucket
+    bucket?.updateRateLimit(response);
+
+    //Update request -> bucket mapping
+    if (bucket != null) {
+      _bucketByRequestRateLimitId.update(
+        request.rateLimitId,
+        (b) => bucket,
+        ifAbsent: () => bucket,
+      );
     }
 
-    if (!request.rateLimit) {
-      return _handle(await _noRateBucket.execute(request));
-    }
-
-    final bucket = _getBucketForRequest(request);
-    return _handle(await bucket.execute(request));
+    return bucket;
   }
 
-  HttpBucket _getBucketForRequest(HttpRequest request) {
-    final reactionsRegexMatch = _bucketReactionsRegexp.firstMatch(request.uri.toString());
-    if (reactionsRegexMatch != null) {
-      final bucketMajorId = reactionsRegexMatch.group(1);
-      final bucketMessageId = reactionsRegexMatch.group(2);
-
-      return _findBucketById("reactions/$bucketMajorId/$bucketMessageId");
-    }
-
-    final commandPermissionRegexMatch = _bucketCommandPermissions.firstMatch(request.uri.toString());
-    if (commandPermissionRegexMatch != null) {
-      final bucketMajorId = commandPermissionRegexMatch.group(1);
-      final bucketMessageId = commandPermissionRegexMatch.group(2);
-
-      return _findBucketById("commands/permissions/$bucketMajorId/$bucketMessageId");
+  Future<HttpResponse> execute(HttpRequest request) async {
+    if (request.auth) {
+      request.headers.addAll({"Authorization": "Bot ${client.token}"});
     }
 
-    final bucketRegexMatch = _bucketRegexp.firstMatch(request.uri.toString());
-    late String bucketId;
-
-    if (bucketRegexMatch == null) {
-      bucketId = request.uri.toString();
-    } else {
-      final bucketName = bucketRegexMatch.group(1);
-      final bucketMajorId = bucketRegexMatch.group(2);
-      bucketId = "${request.method}/$bucketName/$bucketMajorId";
+    HttpBucket? currentBucket = _bucketByRequestRateLimitId.putIfAbsent(request.rateLimitId, () => null);
```

Comment:
Why not just use `_bucketByRequestRateLimitId[request.rateLimitId]`?

If this is intentional, a comment on this line explaining why would be nice